File: bindings/python/samples/pair_recv_sample.py
import zlink
from sample_support import tcp_endpoint, wait_connected


def main():
# --8<-- [start:doc]
    _, endpoint = tcp_endpoint()

    with zlink.create_context() as ctx:
        with zlink.create_pair_socket(ctx) as server:
            with zlink.create_pair_socket(ctx) as client:
                with server.monitor_open(zlink.MonitorEventMask.CONNECTION_READY) as server_monitor:
                    with client.monitor_open(zlink.MonitorEventMask.CONNECTION_READY) as client_monitor:
                        server.bind(endpoint)
                        client.connect(endpoint)
                        wait_connected(server_monitor, client_monitor)

                client.send().message(b"hello-pair").submit()
                received = zlink.create_received()
                if not server.recv_into(received):
                    raise AssertionError("expected pair payload")
                with received:
                    payload = received.to_bytes_list()
                    if payload != [b"hello-pair"]:
                        raise AssertionError(f"unexpected pair payload: {payload!r}")
                print('[pair/recv] send: "hello-pair" → recv: "hello-pair"')
# --8<-- [end:doc]


if __name__ == "__main__":
    main()

File: bindings/javascript/samples/pair_recv_sample.js
// SPDX-License-Identifier: MPL-2.0

'use strict';

const assert = require('node:assert/strict');
const { once } = require('node:events');
const net = require('node:net');
const zlink = require('@zlink-systems/zlink');

async function reservePort() {
  const srv = net.createServer();
  srv.listen(0, '127.0.0.1');
  await once(srv, 'listening');
  const { port } = srv.address();
  await new Promise((resolve, reject) => srv.close((error) => error ? reject(error) : resolve()));
  return port;
}

async function main() {
// --8<-- [start:doc]
  const port = await reservePort();
  const endpoint = `tcp://127.0.0.1:${port}`;
  const ctx = zlink.createContext();
  const server = zlink.createPairSocket(ctx);
  const client = zlink.createPairSocket(ctx);

  try {
    const serverMonitor = server.monitorOpen([zlink.MonitorEventType.ConnectionReady]);
    const clientMonitor = client.monitorOpen([zlink.MonitorEventType.ConnectionReady]);
    try {
      server.bind(endpoint);
      client.connect(endpoint);
      serverMonitor.recv();
      clientMonitor.recv();
    } finally {
      serverMonitor.close();
      clientMonitor.close();
    }

    const sent = 'hello-pair';
    client.send().message(Buffer.from(sent)).submit();

    const received = new zlink.Received();
    server.recv(received);
    try {
      const recv = received.parts[0].data().toString();
      assert.equal(recv, sent);
      console.log(`[pair/recv] send: "${sent}" → recv: "${recv}"`);
    } finally {
      received.close();
    }
  } finally {
    client.close();
    server.close();
    ctx.close();
  }
// --8<-- [end:doc]
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
